crates/matrix-sdk/src/oidc.rs

```diff
@@ -452,7 +427,80 @@ impl Oidc {
             &mut rand::thread_rng(),
         )?;
 
-        let code = (use_authorization_url)(url, validation_data.state.clone()).await?;
+        let state = validation_data.state.clone();
+
+        self.client.inner.oidc_validation_data.write().await.insert(state.clone(), validation_data);
+
+        Ok(OidcAuthenticationUrl { url, state })
+    }
+
+    /// Authorize the given scope by using the Authorization Code flow.
+    ///
+    /// This should be called after the registered client has been restored with
+    /// [`Oidc::restore_registered_client()`] and the URL returned by
+    /// `url_for_authorize_scope_with_authorization_code` has been presented.
+    ///
+    /// # Arguments
+    ///
+    /// * `code` - The code received as part of the redirect URI when login was
+    ///   successful.
+    ///
+    /// * `state` - The unique string that was included in the
+    ///   `OidcAuthenicationUrl` used for authorization.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// use matrix_sdk::{Client, Error};
+    /// # use url::Url;
+    /// # let homeserver = Url::parse("https://example.com").unwrap();
+    /// # let redirect_uri = Url::parse("http://127.0.0.1/oidc").unwrap();
+    /// # let scope = "scope".parse().unwrap();
+    /// # block_on(async {
+    /// # let client = Client::new(homeserver).await?;
+    /// let oidc = client.oidc();
+    ///
+    /// let auth_url = oidc
+    ///     .url_for_authorize_scope_with_authorization_code(scope, redirect_uri)
+    ///     .await?;
+    /// // Open auth_url.url and wait for response at the redirect URI.
+    ///
+    /// # let (code, redirect_state) = (String::new(), String::new());
+    /// if redirect_state != auth_url.state {
+    ///     return Err(Error::UnknownError("wrong state".into()));
+    /// }
+    ///
+    /// // Alternatively, the state can be used as an identifier to handle
+    /// // several authorization requests with a single redirect URI.
+    ///
+    /// let _response =
+    ///     oidc.authorize_scope_with_authorization_code(&auth_url, code).await?;
+    ///
+    /// // The access and refresh tokens can be persisted either from the response,
+    /// // or from one of the `Client::session_tokens*` methods.
+    ///
+    /// // You can now make any request compatible with the requested scope.
+    /// let _me = client.whoami().await?;
+    /// # anyhow::Ok(()) })
+    /// ```
+    ///
+    /// [`ClientErrorCode`]: matrix_sdk::oidc::types::errors::ClientErrorCode
+    pub async fn authorize_scope_with_authorization_code(
```

Comment:
This is probably more appropriate now (though a bit redundant):
```suggestion
    pub async fn finish_authorization_with_authorization_code(
```

Reply:
Would `finish_authorization_with_code` work? Maybe it wouldn't be specific enough?

Reply:
I'd like to keep "authorization code" since that's the name of the flow